### PR TITLE
[1185] More Alt instances

### PR DESCRIPTION
**Pull request by enzief**
Some `Alt` instances for data structures with `def alt` pointing to `plus`.
Implement a part of #1885

### CODE REVIEW

**Comment by enzief** [comment]
More instances to come, I put the PR up for sanity check.

**Comment by enzief** [comment]
@fommil please review.

**Comment by fommil** [comment]
alt all the things

File: core/src/main/scala/scalaz/CorecursiveList.scala [hunk 1]
```diff
@@ -209,6 +209,10 @@ object CorecursiveList extends CorecursiveListInstances {
         }
       }
 
+      @inline
```
Comment (fommil) [inline core/src/main/scala/scalaz/CorecursiveList.scala]: I'm not sure we can inline this

**Comment by enzief** [comment]
What do you mean by "data types that lose information"?

**Comment by fommil** [comment]
consider List for example, we could combine them, pick the left or the right or the non-empty, or the empty. Plus will combine, but Alt will pick the first non-empty.

**Comment by enzief** [comment]
I might need more (not too) formal definition of Alt vs Plus. From @emilypi I has the understanding that `Alt` is the lifted version of `Semigroup`, I never got how exactly `alt` combines the 2 objects though.

P/S: could you please help with the build? I complains that Scala 2.12.0-M4 is missing for one job.

File: core/src/main/scala/scalaz/IList.scala [hunk 2]
```diff
@@ -699,8 +699,10 @@ sealed abstract class IListInstances extends IListInstance0 {
           }
         go(IList(f(a)), INil())
       }
-    }
 
+      def alt[A](a1: => IList[A], a2: => IList[A]): IList[A] =
```
[the remark this answers is not in the document]
Comment (emilypi) [reply core/src/main/scala/scalaz/IList.scala]: `Alt` is expected to act like a `Semigroup`, and `Alternative` is explicitly a lifted `Monoid`. See [here](https://hackage.haskell.org/package/semigroupoids-5.1/docs/src/Data-Functor-Alt.html#line-144)

You can see `Seq` and `NonEmpty` and `IntMap` explicitly refer to their semigroup (and monoid) instances respectively. The only additional requirement for `Alt` and `Alternative` are distributivity over `<$>` and some [additional laws](https://hackage.haskell.org/package/base-4.6.0.1/docs/Control-Applicative.html#t:Alternative) for `Alternative`. `MonadPlus` is intended to represent [choice and failure](https://hackage.haskell.org/package/base-4.6.0.1/docs/src/Control-Monad.html#MonadPlus).

[the remark this answers is not in the document]
Comment (fommil) [reply core/src/main/scala/scalaz/IList.scala]: sure, they are all lifted semigroups but there is some ambiguity around what the expected intuitive behaviour is, especially when it is possible to do one or move of the following, which may all satisfy the laws:

- does it combine or discard non-empty things? (what about when there is no concept of empty?)
- and when discarding, does it pick the left or the right one?

`Alt` seems to favour "discard" whereas `Plus` favours "combine". For some data types, e.g. `Maybe`, both are equivalent, but `List` gives us a glimpse into the possible ambiguities.

[the remark this answers is not in the document]
Comment (fommil) [reply core/src/main/scala/scalaz/IList.scala]: (I really wish Ed Kmett would just write down the laws that differentiate the behaviours... it's the same story for Decidable...)

[the remark this answers is not in the document]
Comment (emilypi) [reply core/src/main/scala/scalaz/IList.scala]: Interesting. Yeah, I agree there's some ambiguity. I look at the list of `MonadPlus` instances, and I see it more often than not referring to explicit good + failing states in things like `STM`, `Maybe`, and `ReadPrec`. I'm not sure what you mean by "discard" - `Alt []` is `(++)`, `Alt (Map k)` is `union`, `Alt Seq` is `mappend` etc. Seems like standard algebraic behavior to me...

[the remark this answers is not in the document]
Comment (enzief) [reply core/src/main/scala/scalaz/IList.scala]: In Haskell it's also concatenation `++`
Please see http://hackage.haskell.org/package/semigroupoids-5.3.1/docs/src/Data.Functor.Alt.html#line-176

[the remark this answers is not in the document]
Comment (fommil) [reply core/src/main/scala/scalaz/IList.scala]: An example of discard would be throwing away a "failing" case. e.g. `io1 <!> io2` if `io1` failed (with an exception) we are discarding that information. Similarly, for `List` `list1 <!> list2` **could** discard the second list and not concatenate.

**Comment by enzief** [comment]
@xuwei-k Could you please have a look? I don't have permission to restart the build.

**Comment by fommil** [comment]
conflicts :sob:

**Comment by emilypi** [comment]
@enzief do you mind sorting out the conflict real quick? Looks good - we can merge this.

**Pull request merged**